src/lib.rs

```diff
@@ -52,6 +54,78 @@ pub trait OpenatDirExt {
 
     /// Create directory and all parents as necessary; no error is returned if directory already exists.
     fn ensure_dir_all<P: openat::AsPath>(&self, p: P, mode: libc::mode_t) -> io::Result<()>;
+
+    /// Create a `FileWriter` which provides a `std::io::BufWriter` and then atomically creates
+    /// the file at the destination, renaming it over an existing one if necessary.
+    fn new_file_writer<'a, P: AsRef<Path>>(
+        &'a self,
+        destname: P,
+        mode: libc::mode_t,
+    ) -> io::Result<FileWriter>;
+
+    /// Atomically create or replace the destination file, calling the provided
+    /// function to generate the contents.  Note that the contents of the
+    /// file will not be explicitly sync'd to disk; if you want to do so you
+    /// need to invoke `writer.flush()?; writer.get_ref().sync_all()` for example.
+    fn write_file_with<P: AsRef<Path>, F, T, E>(
+        &self,
+        destname: P,
+        mode: libc::mode_t,
+        f: F,
+    ) -> Result<T, E>
+    where
+        F: Fn(&mut std::io::BufWriter<std::fs::File>) -> Result<T, E>,
+        E: From<io::Error>,
+    {
+        let mut w = self.new_file_writer(destname, mode)?;
+        match f(&mut w.writer) {
+            Ok(v) => {
+                w.complete()?;
+                Ok(v)
+            }
+            Err(e) => {
+                w.abandon();
+                Err(e)
+            }
+        }
+    }
+
+    /// Like `write_file_with()` but explicitly synchronizes the target to disk.
+    fn write_file_with_sync<P: AsRef<Path>, F, T, E>(
+        &self,
+        destname: P,
+        mode: libc::mode_t,
+        f: F,
+    ) -> Result<T, E>
+    where
+        F: Fn(&mut std::io::BufWriter<std::fs::File>) -> Result<T, E>,
+        E: From<io::Error>,
+    {
+        let mut w = self.new_file_writer(destname, mode)?;
+        match f(&mut w.writer) {
+            Ok(v) => {
+                w.complete_with(|f| {
```

Comment:
Who is meant to do the BufWriter flushing in this codepath? I was expecting the library to do it here right before the `complete_with`.

Reply:
https://github.com/cgwalters/openat-ext/pull/8/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R313 does it, see https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_inner